Utility::Debug: add std::ostream output operator call converter

rune-scape opened the pull request:
Compile errors are much clearer for types without a stream operator, and avoids being an actual catch-all.

This is a bit of a selfish pr, as my code to translate `Debug` `operator<<`s to `std::ostream` `operator<<`s is bugged when there is a catchall for Debug stream `operator<<`s.
Both stream fallbacks use each other. This results in being able to compile `cout << UnprintableClass{}`.
At runtime it gets stuck in recursion, flipflopping between both `operator<<`s. I thought that was pretty interesting.

Please tell me if I made any formatting mistakes.

rune-scape committed:
Utility::Debug: DebugOstreamFallback enable if has ostream operator

mosra commented on src/Corrade/Utility/DebugStl.h:
```diff
@@ -103,9 +103,32 @@ template<class ...Args> Debug& operator<<(Debug& debug, const std::tuple<Args...
 namespace Implementation {
 
+template<typename T>
+typename std::add_lvalue_reference<T>::type declare_lvalue_reference() noexcept;
```
This is basically "like `std::declval<T>()`, but returning a T& instead of T&&", right? :) I think this might be useful in other contexts as well, will put it into `TypeTraits.h` as a public API.

bowling-allie replied:
yes! `std::delcval` declares an rvalue refrence, so it picked up the `operator<<` for `Debug&&` instead of the ones wanted

mosra commented on src/Corrade/Utility/DebugStl.h:
```diff
@@ -103,9 +103,32 @@ template<class ...Args> Debug& operator<<(Debug& debug, const std::tuple<Args...
+
+    public:
+        static constexpr bool value = decltype(check<T>(nullptr))::value;
+};
```
There's a (admittedly weirdly named) [CORRADE_HAS_TYPE(className, typeExpression)](https://doc.magnum.graphics/corrade/TypeTraits_8h.html#a19ce628bb598766db0fffb24c19fc9de) macro that already contains the above functionality, but your variant seems to be a lot simpler than what I stole from somewhere back in the day :) If I fix the macro to have the second parameter variadic (so the comma doesn't break it), I *think* we could then reduce the above to be just this (containing the potentially hard-to-understand template magic in a single place):

```cpp
CORRADE_HAS_TYPE(HasOstreamOperator, typename std::is_same<
    decltype(declareLvalueReference<std::ostream>() << std::declval<C>()),
    std::add_lvalue_reference<std::ostream>::type
>::type)
```

(Also adapting the naming to what Corrade uses for type traits.) I hope I didn't overlook something that would make this impossible tho.

mosra replied:
`CORRADE_HAS_TYPE()` is fixed to allow this in 8d4eb5bc333d8b8e58e5bbbc9946999f17f1648d, and I think it would work with a minor change:

```cpp
CORRADE_HAS_TYPE(HasOstreamOperator, typename std::enable_if<std::is_same<
    decltype(declareLvalueReference<std::ostream>() << std::declval<C>()),
    std::add_lvalue_reference<std::ostream>::type
>::value>::type)
```

I assume you have some way to test the infinite recursion already -- can you confirm? :)

bowling-allie replied:
yes! works perfectly. A very useful macro too

mosra reviewed (commented):
The extremely huge compiler errors on a mismatched operator<< were a pain, so thank you for making everyone's life better :)

> This is a bit of a selfish pr

Why? :) When I saw the notification, I actually thought you'd be adding the inverse operator here -- making `Debug`-printable types work with `std::ostream`. I'm happy to accept that, you'd be definitely not the only one using it.

rune-scape committed:
Merge https://github.com/mosra/corrade into pr88

rune-scape committed:
Move DeclareLvalueReference to TypeTraits.h
simplified HasOstreamOperator by using CORRADE_HAS_TYPE macro
renamed both to magnum schemes

bowling-allie commented:
I figured I should separate this from the feature adding one, but if you want I can add it in.

mosra commented:
Adding the inverse operator would have a good side effect of testing that the above SFINAE restriction actually works and there's no infinite recursion anymore (while this alone doesn't check for that, and so I'm afraid of breaking it by accident).

So, by all means, add it in :)

rune-scape committed:
Utility::Debug: add HasDebugStreamOperator check to DebugOstreamFallback

rune-scape committed:
Utility::Debug: added std::ostream output operator call converter.

bowling-allie commented:
The extra checks are to avoid ambiguous calls

rune-scape committed:
Utility: doc++

mosra commented on src/Corrade/Utility/TypeTraits.h:
```diff
@@ -163,4 +163,10 @@ template<class T> using IsStringLike = std::integral_constant<bool,
+@brief Like @cpp std::declval @ce, but declares an lvalue reference
+*/
+template<typename T>
+typename std::add_lvalue_reference<T>::type DeclareLvalueReference() noexcept;
```
Minor: I'm usually not putting a newline after `template<..>`, keeping the whole signature a single line. (Or generally not adding a newline unless really needed to keep the code vertically terse.)

mosra commented on src/Corrade/Utility/DebugStl.h:
```diff
@@ -130,4 +149,14 @@ CORRADE_UTILITY_EXPORT Debug& operator<<(Debug& debug, Implementation::DebugOstr
 
 }}
 
+template<typename T>
```
One last thing: can you 

- add a short documentation block for this new function,
- mention this in the class documentation so people get to know it exists -- there's a "Printing STL types" section, so maybe extending that a bit ("Printing STL types, interaction with iostreams"?)
- and add a test in `Test/DebugTest.cpp`? There's `ostreamFallback` and `ostreamFallbackPriority`, so something similar (`ostreamDelegate`?) and also similarly checking the priority works for the inverse case, picking the ostream operator over the Debug one -- btw., don't forget to list the two new methods in the `addTests()` so these get properly called

Thank you! :+1:

bowling-allie replied:
yes! wow `ostreamDelegate` is way better than any name I considered

mosra reviewed (commented):
This is great, the diff is so nicely small and tidy that at first I thought I was looking at the previous state with the operator<< not added yet :)

bowling-allie commented:
a problem:
implicit conversions take priority over the operator<<(only because HasOstreamOperator only checks for any possible ostream print operator), meaning Container::Array get printed as a void *

a fix is making `HasOstreamOperatorExact` and `HasOstreamMemberOperatorExact`, and checking like 4 versions of each... but like.. no

rune-scape committed:
Utility::Debug: OstreamDebug::operator<< is now in its own namespace
added documentation
added a snippet

rune-scape committed:
Utility::Debug: more docs

rune-scape committed:
Utility: snippet had narrowing conversion

rune-scape committed:
Utility: DebugTest.cpp: corrected sourceLocation test
adding tests before the sourceLocation changed the line numbers
it was testing

mosra committed:
Utility: whitespace at EOL.

mosra committed:
Utility: this is how it is expected to work.

mosra committed:
Utility: eureka!

mosra commented:
@bowling-allie finally today I got a totally random idea and it seems it works -- or at least made the test pass. Full details in the 3 new commits I just pushed to your branch (I hope that's okay, hehe). 

Can you check with your own code if everything is still behaving as expected and no circular recursion or anything happens again?
